pkg/dispatcher/prometheus.go

```diff
@@ -0,0 +1,98 @@
+package dispatcher
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
+)
+
+type basicAuthRoundTripper struct {
+	username             string
+	password             string
+	originalRoundTripper http.RoundTripper
+}
+
+func (rt *basicAuthRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	req.SetBasicAuth(rt.username, rt.password)
+	return rt.originalRoundTripper.RoundTrip(req)
+}
+
+// PrometheusAPI defines what we expect Prometheus to do in the package
+type PrometheusAPI interface {
+	// Query performs a query for the given time.
+	Query(ctx context.Context, query string, ts time.Time) (model.Value, v1.Warnings, error)
+}
+
+// SaveJobVolumes saves job volumes to a file
+func SaveJobVolumes(jobVolumes map[string]float64, path string) error {
+	bytes, err := yaml.Marshal(jobVolumes)
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(path, bytes, 0644)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// LoadJobVolumes loads job volumes from a file
+// If the file does not exist, return an empty map instead of raising an error.
+func LoadJobVolumes(path string) (map[string]float64, error) {
+	bytes, err := ioutil.ReadFile(path)
+	if err != nil {
+		//keep backwards compatibility
+		if os.IsNotExist(err) {
+			bytes = []byte{}
+		} else {
+			return nil, err
+		}
+	}
+	jobVolumes := map[string]float64{}
+	if err = yaml.Unmarshal(bytes, &jobVolumes); err != nil {
+		return nil, err
+	}
+
+	return jobVolumes, nil
+}
+
+// GetJobVolumesFromPrometheus gets job volumes from a Prometheus server
+func GetJobVolumesFromPrometheus(ctx context.Context, prometheusAPI PrometheusAPI) (map[string]float64, error) {
+	result, warnings, err := prometheusAPI.Query(ctx, `sum(increase(prowjob_state_transitions{state="pending"}[7d])) by (job_name)`, time.Now())
+	if err != nil {
+		return nil, err
+	}
+	if len(warnings) > 0 {
+		logrus.WithField("Warnings", warnings).Warn("Got warnings from Prometheus")
+	}
+
+	vector, ok := result.(model.Vector)
+	if !ok {
+		return nil, fmt.Errorf("returned result from Prometheus cannot be cast to vector")
```

Comment:
please add a type formatter for `result` (`%T`), if we ever hit this someone will be really annoyed at the useless error

Reply:
Done.

Reply:
Done.